crates/voicevox_core_c_api/src/compatible_engine.rs

```diff
@@ -359,6 +359,99 @@ pub unsafe extern "C" fn decode_forward(
+pub unsafe extern "C" fn generate_full_intermediate(
+    length: i64,
+    phoneme_size: i64,
+    margin_width: i64,
+    feature_dim: i64,
+    f0: *mut f32,
+    phoneme: *mut f32,
+    speaker_id: *mut i64,
+    output: *mut f32,
+) -> bool {
+    init_logger_once();
+    assert_aligned(f0);
+    assert_aligned(phoneme);
+    assert_aligned(speaker_id);
+    assert_aligned(output);
+    let length = length as usize;
+    let phoneme_size = phoneme_size as usize;
+    let margin_width = margin_width as usize;
+    let feature_dim = feature_dim as usize;
+    let synthesizer = &*lock_synthesizer();
+    let result = ensure_initialized!(synthesizer).generate_full_intermediate(
+        length,
+        phoneme_size,
+        // SAFETY: The safety contract must be upheld by the caller.
+        unsafe { std::slice::from_raw_parts(f0, length) },
+        unsafe { std::slice::from_raw_parts(phoneme, phoneme_size * length) },
+        StyleId::new(unsafe { *speaker_id as u32 }),
+    );
+    match result {
+        Ok(output_arr) => {
+            // SAFETY: The safety contract must be upheld by the caller.
+            let output_slice = unsafe {
+                std::slice::from_raw_parts_mut(output, (length + 2 * margin_width) * feature_dim)
+            };
+            output_slice.clone_from_slice(&output_arr.into_raw_vec());
+            true
+        }
+        Err(err) => {
+            set_message(&format!("{err}"));
+            false
+        }
+    }
+}
+
+/// # Safety
+///
+/// - `audio_feature`はRustの`&[f32; (length * feature_dim) as usize]`として解釈できなければならない。
+/// - `speaker_id`はRustの`&[i64; 1]`として解釈できなければならない。
+/// - `output`はRustの`&mut [f32; length as usize * 256]`として解釈できなければならない。
+#[unsafe(no_mangle)] // SAFETY: voicevox_core_c_apiを構成するライブラリの中に、これと同名のシンボルは存在しない
+pub unsafe extern "C" fn render_audio_segment(
+    length: i64,
+    feature_dim: i64,
+    audio_feature: *mut f32,
+    speaker_id: *mut i64,
+    output: *mut f32,
+) -> bool {
+    init_logger_once();
+    assert_aligned(audio_feature);
+    assert_aligned(speaker_id);
+    let length = length as usize;
+    let feature_dim = feature_dim as usize;
+    let synthesizer = &*lock_synthesizer();
+    // SAFETY: The safety contract must be upheld by the caller.
+    let audio_feature_vec =
+        unsafe { std::slice::from_raw_parts(audio_feature, length * feature_dim) };
+    let result = ensure_initialized!(synthesizer).render_audio_segment(
+        ndarray::arr1(audio_feature_vec)
+            .into_shape([length, feature_dim])
+            .unwrap(),
+        StyleId::new(unsafe { *speaker_id as u32 }),
+    );
+    match result {
+        Ok(output_arr) => {
+            // SAFETY: The safety contract must be upheld by the caller.
+            let output_slice = unsafe { std::slice::from_raw_parts_mut(output, length * 256) };
+            output_slice.clone_from_slice(&output_arr.to_vec());
```

Comment:
出力側についても:

```suggestion
            let output_arr = output_arr.as_standard_layout();
            // SAFETY: The safety contract must be upheld by the caller.
            unsafe {
                output_arr
                    .as_ptr()
                    .copy_to_nonoverlapping(output, length * 256);
            }
```

（`.as_standard_layout()`については`.is_standard_layout()`でのアサートでも別に問題ないとは思いますが、`as_standard_layout`は元々"standard"なやつをコピーしたりはしないので、よく考えずに済むこっちの方がよいかなと思いました）

Reply:
> [追記] あと、これをやった場合、やらなくてもいいとは思いますが、この関数の"Safety"の&mut [f32; length as usize * 256]は&mut [MaybeUninit<f32>; length as usize * 256]にできると思います。というのもnumpy.zerosじゃなくて[numpy.empty](https://numpy.org/doc/stable/reference/generated/numpy.empty.html)が一応使えそうなので…

これの違いが判らなかったんですが、`&mut [MaybeUninit<f32>; length as usize * 256]` は `&mut [f32; length as usize * 256]` と比べて何が保証されていないんですか？uninitな値を読もうとするとおかしくなることがあるんでしょうか（C++でいうundefined behavior的な...)

Reply:
> https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#:~:text=any%20fixed%20bit-,pattern%3A,-use%20std%3A%3Amem

Undefined Behaviorですね。

Reply:
まずUB (Undefined Behavior)の何が駄目なのかというと、[最悪タイムトラベル的な挙動が起きうる](https://cpplover.blogspot.com/2014/06/old-new-thing.html)ことかなーと。

そこで例えばCの`float*`をRustの`&mut [f32; N]`として読むことを考えます。まず

- `&mut`は中身を「専有」しなければならない。つまりC側とかは中身に一切触ってはならない
- ポインタはalignedじゃないといけない

は守る必要があると思います。この二つはタイムトラベル的なことが起きうると思います (起きるかどうかわかりませんし、具体例を出すのが難しいですが)。

それで「未初期化であってはならない」ですが、FFI越しに守る必要があるかというとちょっと微妙かもしれません。というのもコンパイル単位越しにタイムトラベルが起きるかは多分微妙だからです。ただ上の二つと線引きすべきかというとそれもまた微妙かと思いますし、あらゆるアーキテクチャとプラットフォーム上でも大丈夫なのかについても「わからない」とした方がいいんじゃないかなと個人的には思ってます。

Reply:
あ、名無し。さんが出してくれた範囲に書いてましたね。普通に起きうりますねタイムトラベルが。